### PR TITLE
Implement viewer CID opt-in

### LINKED ISSUE #10518: Viewer CID pub opt-in
Allow pub to specify a special meta tag in `<head>` to opt-in Viewer CID API.

```
<meta name="amp-google-client-id-api" content="googleanalytics">
```

The above meta tag translates to
if the current Viewer is under google origin, use Viewer CID API to retrieve CID for googleanalytics' scope (AMP_ECID_GOOGLE) .

### PR DESCRIPTION
**Pull request by lannka**
Closes #10518

### CODE REVIEW

File: src/service/viewer-cid-api.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+import {dict} from '../utils/object';
+
+const GOOGLE_CLIENT_ID_API_META_NAME = 'amp-google-client-id-api';
+const GOOGLE_VIEWER_ORIGIN_REGEX =
+    /(^|\.)google\.(com?|[a-z]{2}|com?\.[a-z]{2}|cat)$/;
+const CID_API_SCOPE_WHITELIST = {
+  'googleanalytics': 'AMP_ECID_GOOGLE',
+};
+
+export class ViewerCidApi {
```
Comment (cramforce) [inline src/service/viewer-cid-api.js]: Sprinkle some docs.

File: src/service/viewer-cid-api.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  /**
+   * @return {!Promise<boolean>}
+   */
+  shouldGetScopedCid(scope) {
+    if (!this.viewer_.hasCapability('cid')
+        || !this.isScopeOptedInForCidApi_(scope)) {
+      return Promise.resolve(false);
+    }
+
+    return this.viewer_.getViewerOrigin().then(origin => {
```
Comment (cramforce) [inline src/service/viewer-cid-api.js]: Can't this `viewer.isTrustedViewer`

Comment (cramforce) [inline src/service/viewer-cid-api.js]: I see you want it Google specific, right?

File: src/service/viewer-cid-api.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    return this.viewer_.sendMessageAwaitResponse('cid', dict({
+      scope,
+      clientIdApi: true,
+    }));
+  }
+
+  /**
+   * @return {!Promise<boolean>}
+   */
+  shouldGetScopedCid(scope) {
```
Comment (cramforce) [inline src/service/viewer-cid-api.js]: I think we should always get a scoped CID when the viewer supports it (and is trusted). The `sScopeOptedInForCidApi_` should determine whether we request usage of the API.

Comment (lannka) [inline src/service/viewer-cid-api.js]: done.

File: src/service/viewer-cid-api.js (hunk 1 above)
Comment (lannka) [inline src/service/viewer-cid-api.js]: done

**Comment by lannka** [comment]
PTAL

File: src/service/viewer-cid-api.js [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+import {Services} from '../services';
+import {dict} from '../utils/object';
+
+const GOOGLE_CLIENT_ID_API_META_NAME = 'amp-google-client-id-api';
+const CID_API_SCOPE_WHITELIST = {
+  'googleanalytics': 'AMP_ECID_GOOGLE',
+};
+
+/**
+ * Exposes CID API if provided by the Viewer.
```
Comment (cramforce) [inline src/service/viewer-cid-api.js]: Also rename file to `GoogleViewerCidApi`

File: src/service/viewer-cid-api.js [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+  /**
+   * Returns scoped CID retrieved from the Viewer.
+   * @param {!string} scope
+   * @return {!Promise<?string>}
+   */
+  getScopedCid(scope) {
+    return this.shouldUseCidApi_(scope).then(clientIdApi => {
+      return this.viewer_.sendMessageAwaitResponse('cid', dict({
+        scope,
```
Comment (cramforce) [inline src/service/viewer-cid-api.js]: See lint error. Must be `'scope': scope`

File: src/service/viewer-cid-api.js [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    this.ampdoc_ = ampdoc;
+    this.viewer_ = Services.viewerForDoc(this.ampdoc_);
+  }
+
+  /**
+   * Returns true if Viewer supports CID API.
+   * @returns {boolean}
+   */
+  isSupported() {
+    return this.viewer_.hasCapability('cid');
```
Comment (cramforce) [inline src/service/viewer-cid-api.js]: This also needs to be guarded by `isTrustedViewer`.

`shouldUseCidApi_` can then just be `isScopeOptedInForCidApi_`

File: src/service/viewer-cid-api.js (hunk 4 above)
Comment (lannka) [inline src/service/viewer-cid-api.js]: shall we leave it for other Viewer integration?

File: src/service/viewer-cid-api.js (hunk 6 above)
Comment (lannka) [inline src/service/viewer-cid-api.js]: good point

File: src/service/viewer-cid-api.js (hunk 5 above)
Comment (lannka) [inline src/service/viewer-cid-api.js]: done

**Pull request merged**